File: src/main/java/ru/nucodelabs/files/sonet/SonetImportUtils.java
package ru.nucodelabs.files.sonet;

import java.io.File;
import java.util.ArrayList;
import java.util.Locale;
import java.util.Scanner;

final class SonetImportUtils {
    private SonetImportUtils() {
    } // чтобы нельзя было создать экземпляр класса Sonet

    static STTFile readSTT(File file) throws Exception {
        STTFile res = new STTFile();
        Scanner sc = new Scanner(file).useLocale(Locale.US);

        ArrayList<ArrayList<Double>> numbers = columnReader(sc, new STTFile().getColumnCnt());

        res.getAB_2().addAll(
                numbers.stream().map(s -> s.get(0)).toList());
        res.getMN_2().addAll(
                numbers.stream().map(s -> s.get(1)).toList());

        sc.close();
        res.setFile(file);
        return res;
    }

    static EXPFile readEXP(File file) throws Exception {
        EXPFile res = new EXPFile();
        Scanner sc = new Scanner(file, "Cp866").useLocale(Locale.US);
        res.setSTTFileName(sc.nextLine());
        readPassport(sc, res);
        ArrayList<ArrayList<Double>> numbers = null;
        if (sc.hasNext("\\$")) {
            sc.next();
            numbers = columnReader(sc, new EXPFile().getColumnCnt());
        }
        assert numbers != null;

        res.getAmperage().addAll(
                numbers.stream().map(s -> s.get(0)).toList());
        res.getVoltage().addAll(
                numbers.stream().map(s -> s.get(1)).toList());
        res.getResistanceApparent().addAll(
                numbers.stream().map(s -> s.get(2)).toList());
        res.getErrorResistanceApparent().addAll(
                numbers.stream().map(s -> s.get(3)).toList());
        res.getPolarizationApparent().addAll(
                numbers.stream().map(s -> s.get(4)).toList());
        res.getErrorPolarizationApparent().addAll(
                numbers.stream().map(s -> s.get(5)).toList());

        sc.close();
        res.setFile(file);
        return res;
    }

    static MODFile readMOD(File file) throws Exception {
        MODFile res = new MODFile();
        Scanner sc = new Scanner(file).useLocale(Locale.US);

        ArrayList<ArrayList<Double>> numbers = columnReader(sc, new MODFile().getColumnCnt());

        res.getResistance().addAll(
                numbers.stream().map(s -> s.get(0)).toList());
        res.getPower().addAll(
                numbers.stream().map(s -> s.get(1)).toList());
        res.getPolarization().addAll(
                numbers.stream().map(s -> s.get(2)).toList());

        sc.close();
        res.setFile(file);
        return res;
    }

    private static void readPassport(Scanner sc, EXPFile res) {
        ArrayList<String> strList = new ArrayList<>();
        while (sc.hasNextLine() && !sc.hasNext("\\$") && strList.size() < 6) {
            strList.add(sc.nextLine());
        }
        if (strList.size() > 0) res.setNumber(strList.get(0));
        if (strList.size() > 1) res.setDate(strList.get(1));
        if (strList.size() > 2) res.setWeather(strList.get(2));
        if (strList.size() > 3) res.setOperator(strList.get(3));
        if (strList.size() > 4) res.setInterpreter(strList.get(4));
        if (strList.size() > 5) res.setChecked(strList.get(5));
    }

    private static ArrayList<ArrayList<Double>> columnReader(Scanner sc, int colCnt) {
        ArrayList<ArrayList<Double>> res = new ArrayList<>();
        while (sc.hasNextLine() && !sc.hasNext("\\$") && !sc.hasNext("-1.0")) {
            String str = sc.nextLine();
            if (str.isBlank()) {
                continue;
            }
            Scanner strSc = new Scanner(str).useLocale(Locale.US);
            ArrayList<Double> numList = new ArrayList<>();
            for (int i = 0; i < colCnt; i++) {
                if (strSc.hasNext()) {
                    numList.add(strSc.nextDouble());
                } else {
                    numList.add(0d);
                }
            }
            res.add(numList);
            strSc.close();
        }
        return res;
    }
}

File: src/main/java/ru/nucodelabs/files/sonet/SonetImportUtils.kt
package ru.nucodelabs.files.sonet

import java.io.File
import java.util.*

internal object SonetImportUtils {
    @JvmStatic
    @Throws(Exception::class)
    fun readSTT(file: File): STTFile {
        val res = STTFile()
        Scanner(file).useLocale(Locale.US).use { sc ->
            val numbers = columnReader(sc, STTFile().columnCnt)
            res.AB_2.addAll(numbers.map { it[0] })
            res.MN_2.addAll(numbers.map { it[1] })
        }
        res.file = file
        return res
    }

    @JvmStatic
    @Throws(Exception::class)
    fun readEXP(file: File): EXPFile {
        val res = EXPFile()
        Scanner(file, "Cp866").useLocale(Locale.US).use { sc ->
            res.STTFileName = sc.nextLine()
            readPassport(sc, res)
            val numbers = if (sc.hasNext("\\$") ) {
                sc.next(); columnReader(sc, EXPFile().columnCnt)
            } else null
            requireNotNull(numbers)
            res.amperage.addAll(numbers.map { it[0] })
            res.voltage.addAll(numbers.map { it[1] })
            res.resistanceApparent.addAll(numbers.map { it[2] })
            res.errorResistanceApparent.addAll(numbers.map { it[3] })
            res.polarizationApparent.addAll(numbers.map { it[4] })
            res.errorPolarizationApparent.addAll(numbers.map { it[5] })
        }
        res.file = file
        return res
    }

    @JvmStatic
    @Throws(Exception::class)
    fun readMOD(file: File): MODFile {
        val res = MODFile()
        Scanner(file).useLocale(Locale.US).use { sc ->
            val numbers = columnReader(sc, MODFile().columnCnt)
            res.resistance.addAll(numbers.map { it[0] })
            res.power.addAll(numbers.map { it[1] })
            res.polarization.addAll(numbers.map { it[2] })
        }
        res.file = file
        return res
    }

    private fun readPassport(sc: Scanner, res: EXPFile) {
        val strList = ArrayList<String>()
        while (sc.hasNextLine() && !sc.hasNext("\\$") && strList.size < 6) {
            strList.add(sc.nextLine())
        }
        if (strList.size > 0) res.number = strList[0]
        if (strList.size > 1) res.date = strList[1]
        if (strList.size > 2) res.weather = strList[2]
        if (strList.size > 3) res.operator = strList[3]
        if (strList.size > 4) res.interpreter = strList[4]
        if (strList.size > 5) res.checked = strList[5]
    }

    private fun columnReader(sc: Scanner, colCnt: Int): MutableList<MutableList<Double>> {
        val res = mutableListOf<MutableList<Double>>()
        while (sc.hasNextLine() && !sc.hasNext("\\$") && !sc.hasNext("-1.0")) {
            val line = sc.nextLine()
            if (line.isBlank()) continue
            Scanner(line).useLocale(Locale.US).use { rowScanner ->
                val row = MutableList(colCnt) { if (rowScanner.hasNextDouble()) rowScanner.nextDouble() else 0.0 }
                res.add(row)
            }
        }
        return res
    }
}
